### PR TITLE
Add isolate samples and upgrade to mono_repo 6.0.0

### PR DESCRIPTION
**Pull request by johnpryan**

### CODE REVIEW

File: isolates/README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Additional samples for [Concurrency in Dart][concurrency-in-dart].
+
+- `bin/send_and_receive.dart`: Demonstrates how to spawn an isolate and
+  communicate with it using a SendPort. The spawned isolate first sends a
+  SendPort for the main isolate to communicate with.
+- `bin/long_running_isolate.dart`: Similar to `send_and_receive.dart`, except
+  that the spawned isolate responds to incoming messages from the main isolate,
+  sending a response until `null` is sent over the SendPort.
+
+[concurrency-in-dart]: https://dart.dev/guides/language/concurrency
```
Comment (parlough) [inline isolates/README.md:10]: How about adding some useful doc links for referencing?

```suggestion
- `bin/send_and_receive.dart`: Demonstrates how to spawn an [`Isolate`][] and
  communicate with it using a [`SendPort`][]. The spawned isolate first sends a
  SendPort for the main isolate to communicate with.
- `bin/long_running_isolate.dart`: Similar to `send_and_receive.dart`, except
  that the spawned isolate responds to incoming messages from the main isolate,
  sending a response until `null` is sent over the [`SendPort`][].

[concurrency-in-dart]: https://dart.dev/guides/language/concurrency
[`Isolate`]: https://api.dart.dev/stable/dart-isolate/Isolate-class.html
[`SendPort`]: https://api.dart.dev/stable/dart-isolate/SendPort-class.html
```

File: isolates/pubspec.lock [hunk 2]
```diff
@@ -0,0 +1,341 @@
+# Generated by pub
```
Comment (kevmoo) [inline isolates/pubspec.lock:1]: Let's ignore this, too. Just creates noise as things are upgraded.

File: isolates/bin/long_running_isolate.dart [hunk 3]
```diff
@@ -0,0 +1,88 @@
+// Spawn an isolate, read multiple files, send their contents to the spawned
+// isolate, and wait for the parsed JSON.
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:isolate';
+
+final files = [
+  File('json_01.json'),
```
Comment (kevmoo) [inline isolates/bin/long_running_isolate.dart:9]: 🤷  – maybe make this `String` and create the `File` object below? So we don't have File instances hanging around...

File: isolates/pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+name: isolates
+description: A sample command-line application.
+version: 1.0.0
+
+environment:
+  sdk: '>=2.14.1 <3.0.0'
```
Comment (kevmoo) [inline isolates/pubspec.yaml:6]: Is this a 2.15 feature? Should that be the lower bound?

File: isolates/mono_pkg.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+sdk:
+  - dev
+  - stable
```
Comment (kevmoo) [inline isolates/mono_pkg.yaml:3]: I usually put this as an EXPLICIT SDK version – matching the lower SDK constraint in the pubspec

File: isolates/bin/send_and_receive.dart [hunk 6]
```diff
@@ -0,0 +1,44 @@
+// Read the file, spawn an isolate, send the file contents to the spawned
+// isolate, and wait for the parsed JSON.
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:isolate';
+
+final filename = 'json_01.json';
```
Comment (parlough) [inline isolates/bin/send_and_receive.dart:8]:
```suggestion
const filename = 'json_01.json';
```

File: isolates/bin/long_running_isolate.dart [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  File('json_02.json'),
+  File('json_03.json'),
+];
+
+// Reads multiple files and parses them on a separate isolate.
+Future<void> main() async {
```
Comment (kevmoo) [inline isolates/bin/long_running_isolate.dart:15]: Might be nice to have @lrhn or @mkustermann review these!

File: isolates/bin/send_and_receive.dart [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+Future<void> main() async {
```
Comment (kevmoo) [inline isolates/bin/send_and_receive.dart:10]: Might be nice to have @lrhn or @mkustermann review this!

File: isolates/pubspec.yaml (hunk 4 above)
Comment (johnpryan) [reply isolates/pubspec.yaml:6]: yes, good catch!

File: isolates/mono_pkg.yaml (hunk 5 above)
Comment (johnpryan) [reply isolates/mono_pkg.yaml:3]: Ah, makes sense. I'll update this

File: isolates/bin/long_running_isolate.dart [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+      // This is the first message, assign the sendPort
+      sendPort = message;
+    } else if (message is Map<String, dynamic>) {
+      // This is a response, add it to the Stream returned by this function.
+      yield message;
+    }
+
+    // Send the next JSON string to be parsed.
+    if (i < jsonStrings.length) {
+      sendPort.send(jsonStrings[i++]);
```
Comment (parlough) [inline isolates/bin/long_running_isolate.dart:58]: I think this inline postfix increment can cause some temporary confusion for some developers reading this. What do you think about splitting it?

```suggestion
      sendPort.send(jsonStrings[i]);
      i++;
```

or

```suggestion
      sendPort.send(jsonStrings[i]);
      i += 1;
```

**Review by parlough — APPROVED** [review]
The examples and corresponding comments look great. Some optional, minor suggestions.

File: isolates/bin/long_running_isolate.dart [hunk 10]
```diff
@@ -0,0 +1,88 @@
+// Spawn an isolate, read multiple files, send their contents to the spawned
+// isolate, and wait for the parsed JSON.
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:isolate';
+
+final fileNames = [
```
Comment (parlough) [inline isolates/bin/long_running_isolate.dart:8]:
```suggestion
const filenames = [
```

File: isolates/bin/long_running_isolate.dart [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+  // Use that data.
+  await for (final jsonData in _parseJsonStrings(fileData)) {
+    print('Received JSON with ${jsonData.length} keys');
+  }
+}
+
+// Reads multiple files simultaneously using Future.wait().
+Future<List<String>> _readFiles() async {
+  return await Future.wait([
+    ...fileNames.map((f) => File(f)).map((file) => file.readAsString()),
```
Comment (parlough) [inline isolates/bin/long_running_isolate.dart:28]:
```suggestion
    ...filenames.map((f) => File(f)).map((file) => file.readAsString()),
```

File: isolates/bin/long_running_isolate.dart (hunk 7 above)
Comment (lrhn) [reply isolates/bin/long_running_isolate.dart:15]: I'd make this `void main() async {`. There is no reason to suggest that anyone will ever see the result of calling `main`.

File: isolates/bin/long_running_isolate.dart [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+  final p = ReceivePort();
+  await Isolate.spawn(_receiveAndParseJsonService, p.sendPort);
+
+  // Send each JSON string in order, waiting for the previous result before
+  // sending the next JSON string.
+  var i = 0;
+
+  // The port used to send JSON strings to be parsed.
+  // This is the first message sent by the spawned isolate.
+  late final SendPort sendPort;
```
Comment (lrhn) [inline isolates/bin/long_running_isolate.dart:44]: I'd personally avoid using `late`, but using `await for` on a heterogenous stream makes it tricky.
This is ... OK.
I'd probably use [`StreamQueue` from `package:async`](https://pub.dev/documentation/async/latest/async/StreamQueue-class.html) instead, it's more readable:
```dart
  var events = StreamQueue<dynamic>(p);
  SendPort sendPort = await events.next();
  for (var jsonString in jsonStrings) {
    sendPort.send(jsonString);
    Map<String, dynamic> message = await events.next();
    yield message;
  }
  sendPort.send(null);
  await events.cancel();
```

File: isolates/bin/long_running_isolate.dart [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  await for (final message in rp) {
+    if (message is String) {
+      p.send(jsonDecode(message));
+    } else if (message == null) {
+      // Exit if the main isolate sends a null message, indicating there are no
+      // more strings to parse.
+      break;
+    }
+  }
+  print('Spawned isolate finished.');
```
Comment (lrhn) [inline isolates/bin/long_running_isolate.dart:86]: Do we want a symmetric "Spawned isolate started" a the start?

File: isolates/bin/send_and_receive.dart [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  // Read some data.
+  final fileData = await File(filename).readAsString();
+  final jsonData = await _spawnIsolateAndSendJson(fileData);
+
+  // Use that data.
+  print('Received JSON with ${jsonData.length} keys');
+}
+
+// Sends a JSON string to the spawned isolate to be parsed.
+Future<Map<String, dynamic>> _spawnIsolateAndSendJson(String jsonString) async {
```
Comment (lrhn) [inline isolates/bin/send_and_receive.dart:20]: This is unnecessarily complicated when we know we only want a single computation.
Send the string as part of the original message, whether typed or untyped, either is fine.

So:
```dart
Future<Map<String, dynamic>> _spawnIsolateAndSendJson(String jsonString) async {
  final p = ReceivePort();
  await Isolate.spawn(_receiveAndParseJson, [p.sendPort, jsonString]);
  return (await p.first) as Map<String, dynamic>;
}

void _receiveAndParseJson(List<dynamic> args) {
  SendPort responsePort = args[0];
  String jsonString = args[1];
  var result = jsonDecode(jsonString)
  isolate.exit(responsePort, result);
}
```

File: isolates/bin/long_running_isolate.dart [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    }
+  }
+}
+
+// The function runs on the spawned isolate. Waits for incoming strings and
+// sends the decoded JSON using the SendPort.
+Future<void> _receiveAndParseJsonService(SendPort p) async {
+  // Send a SendPort to the main isolate so that it can send JSON strings to
+  // this isolate.
+  final rp = ReceivePort();
```
Comment (lrhn) [inline isolates/bin/long_running_isolate.dart:73]: Avoid abbreviations, so maybe call it "commandPort"?

File: isolates/bin/send_and_receive.dart (hunk 14 above)
Comment (mit-mit) [reply isolates/bin/send_and_receive.dart:20]: I'd suggest moving the reading of the file to the worker isolate, so something like this:

```dart
Future<void> main() async {
  // Read some data.
  final filename = 'json_01.json';
  final jsonData = await _spawnIsolateAndSendJson(filename);

  // Use that data.
  print('Received JSON with ${jsonData.length} keys');
}

Future<Map<String, dynamic>> _spawnIsolateAndSendJson(String fileName) async {
  final p = ReceivePort();
  await Isolate.spawn(_receiveAndParseJson, [p.sendPort, fileName]);
  return (await p.first) as Map<String, dynamic>;
}

void _receiveAndParseJson(List<dynamic> args) async {
  SendPort responsePort = args[0];
  String fileName = args[1];

  final fileData = await File(fileName).readAsString();
  var result = jsonDecode(fileData);
  Isolate.exit(responsePort, result);
}
```

File: isolates/bin/long_running_isolate.dart (hunk 11 above)
Comment (johnpryan) [reply isolates/bin/long_running_isolate.dart:28]: `filename` is more commonly used I think

File: isolates/bin/send_and_receive.dart (hunk 14 above)
Comment (mit-mit) [reply isolates/bin/send_and_receive.dart:20]: Oh, and if so, maybe rename `_receiveAndParseJson` to `_retrieveAndParseJson `

**Comment by mit-mit** [comment]
Hey @johnpryan, would you mind resolving comments that you have already addressed?

File: isolates/bin/long_running_isolate.dart (hunk 9 above)
Comment (johnpryan) [reply isolates/bin/long_running_isolate.dart:58]: Both seem okay, but I went with Lasse's `StreamQueue` idea instead.

File: isolates/bin/send_and_receive.dart (hunk 14 above)
Comment (johnpryan) [reply isolates/bin/send_and_receive.dart:20]: Since it's no longer receiving events explicitly, maybe `_readAndParseJson`?

**Comment by johnpryan** [comment]
Just pushed up 2cbb16d, which should resolve those.

File: isolates/bin/send_and_receive.dart (hunk 14 above)
Comment (johnpryan) [reply isolates/bin/send_and_receive.dart]: Resolved in 2cbb16d

File: isolates/bin/long_running_isolate.dart (hunk 12 above)
Comment (johnpryan) [reply isolates/bin/long_running_isolate.dart]: That's much better, added in 2cbb16d

File: isolates/bin/send_and_receive.dart (hunk 8 above)
Comment (mit-mit) [reply isolates/bin/send_and_receive.dart:14]: Resolving this as Lasse reviewed it

**Review by mit-mit — APPROVED** [review]

**Pull request merged**